### DISCUSSION TITLE
There is no active transaction error in PHPUnit

### DISCUSSION THREAD

**Discussion by rome2o** [laravel/framework#38260] [Q&A]
- Laravel Version: 8.12
- PHP Version: 8.0.1
- Database Driver & Version:
MySQL 8.0.22

### Description:
We are writing SQL test cases for our application and are using DatabaseTransactions in our BaseTestCase.php

The setUp function looks like this

```
    protected function setUp(): void
    {
        parent::setUp();
    }
```

A sample test case is as the following

```
    public function product_with_required_quantitiy_is_returned_in_service_code()
    {
            \App\Models\Inventory::create([
                'product_variant_id'=> $id,
                'warehouse_id'=> $warehouse,
                'qty_onhand'=> $qty,
                'qty_available'=> $qty,
            ]);
            $response = $this->call('POST', route('ecommerce.checkout'), $payload);
            $response
            ->assertStatus(200)
            ->assertJson(fn (AssertableJson $json) => $json->has('rates', 7));

            $data = $response->json();
            $codeDetails = (new \App\Models\ServiceCode($data['rates'][0]['service_code']))->getDetails();
            $this->assertFalse($codeDetails['makeRequest']);
    }
```

However upon execution of the test cases we get the following error

  There is no active transaction

```
  at vendor/laravel/framework/src/Illuminate/Database/Concerns/ManagesTransactions.php:279
    275▕      */
    276▕     protected function performRollBack($toLevel)
    277▕     {
    278▕         if ($toLevel == 0) {
  ➜ 279▕             $this->getPdo()->rollBack();
    280▕         } elseif ($this->queryGrammar->supportsSavepoints()) {
    281▕             $this->getPdo()->exec(
    282▕                 $this->queryGrammar->compileSavepointRollBack('trans'.($toLevel + 1))
    283▕             );
```

Any help would be much appreciated

**Comment by volkv** [comment]
same here... but is was fine like a week ago. trying to understand what has changed

**Reply by volkv** [reply]
I found out in my case it was because of `->truncate()` statement inside test-related code

**Reply by AshishDhamalaAD** [reply]
Thanks, it worked for me.

**Comment by geidelguerra** [comment]
I can confirm that a `truncate()` statement inside a test code causes this error.

**Comment by ChrisSantiago82** [comment]
Is there a workaround to this?

**Reply by SjorsO** [reply]
Instead of using truncate, use a delete query without a where clause, for example: `User::query()->delete()`.

This [StackOverflow post](https://stackoverflow.com/a/66864078/7032830) explains why truncate causes this error.

**Reply by sclausen** [reply]
> Instead of using truncate, use a delete query without a where clause, for example: `User::query()->delete()`.
> 
> This [StackOverflow post](https://stackoverflow.com/a/66864078/7032830) explains why truncate causes this error.

Since I want to reset the autoincrement too, which truncate would have done, I have to do `DB::statement('ALTER TABLE tbl_name AUTO_INCREMENT = 1;');` and am now getting the `There is no active transaction` error again. 😅 It's jinxed!

**Reply by chiragpatel2595** [reply]
@sclausen Did you solve the issue?

**Comment by patriziotomato** [comment]
See also:
https://laravel.com/docs/9.x/database#implicit-commits-in-transactions
